File: backend/app/schemas/watchlist.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class WatchlistItemOut(BaseModel):
    symbol: str
    exchange: str
    name: str
    sort_order: int
    vt_symbol: str
    tf_symbol: str
    # 行情（可选 enrich）
    last_price: float | None = None
    change_pct: float | None = None
    turnover_rate: float | None = None
    volume: float | None = None
    amount: float | None = None
    volume_ratio: float | None = None
    industry: str = ""
    suspended: bool = False


class WatchlistAddRequest(BaseModel):
    symbol: str = Field(description="支持 600519.SSE / SHSE.600519 / 600519")
    name: str = ""
    exchange: str | None = None


class WatchlistReorderRequest(BaseModel):
    items: list[str] = Field(description="vt_symbol 列表，按期望顺序")


class GroupsReorderRequest(BaseModel):
    group_ids: list[str] = Field(min_length=1, description="分组 id 期望顺序")


class GroupOut(BaseModel):
    id: str
    name: str
    sort_order: int


class GroupCreate(BaseModel):
    name: str = Field(min_length=1, max_length=40)


class GroupRename(BaseModel):
    name: str = Field(min_length=1, max_length=40)


class GroupMemberRequest(BaseModel):
    symbol: str
    exchange: str | None = None


class GroupMembersBatchRequest(BaseModel):
    symbols: list[str] = Field(min_length=1, max_length=100)
    action: Literal["add", "remove"]


class GroupMembersBatchError(BaseModel):
    symbol: str
    detail: str


class GroupMembersBatchOut(BaseModel):
    ok: bool = True
    action: Literal["add", "remove"]
    added: int = 0
    removed: int = 0
    skipped: int = 0
    errors: list[GroupMembersBatchError] = Field(default_factory=list)


class QuoteOut(BaseModel):
    symbol: str
    exchange: str
    vt_symbol: str
    tf_symbol: str
    name: str = ""
    last_price: float = 0.0
    change_pct: float = 0.0
    turnover_rate: float = 0.0
    volume: float = 0.0
    amount: float = 0.0
    amplitude: float = 0.0
    volume_ratio: float = 0.0
    industry: str = ""


class BarOut(BaseModel):
    datetime: str
    open: float
    high: float
    low: float
    close: float
    volume: float
    turnover: float


class BarsResponse(BaseModel):
    symbol: str
    exchange: str
    vt_symbol: str
    interval: str
    bars: list[BarOut]


class StrategySignalRow(BaseModel):
    vt_symbol: str
    name: str = ""
    last_price: float | None = None
    change_pct: float | None = None
    signal: str = "na"
    signal_label: str = "—"
    signal_date: str | None = None
    strength: float | None = None
    reason_summary: str = ""
    ref_buy_price: float | None = None
    ref_sell_price: float | None = None
    ma_gap_pct: float | None = None
    volume_ratio_5d: float | None = None
    bar_as_of: str | None = None
    updated_at: str | None = None


class StrategyPositionRow(BaseModel):
    vt_symbol: str
    name: str = ""
    cost_price: float
    volume: int
    buy_date: str
    notes: str = ""
    source: str = "manual"
    plan_pct: float | None = None
    last_price: float | None = None
    market_value: float | None = None
    unrealized_pnl: float | None = None
    unrealized_pnl_pct: float | None = None
    t1_locked: bool = False
    exit_signal: str = "na"
    exit_signal_label: str = "—"
    ref_sell_price: float | None = None
    reason_summary: str = ""
    risk_tags: list[str] = Field(default_factory=list)
    risk_primary: str = ""
    off_plan: bool = False


class PlanSymbolStatus(BaseModel):
    vt_symbol: str
    name: str = ""
    in_watchlist: bool = False
    in_position: bool = False


class RiskSummaryOut(BaseModel):
    total_capital: float | None = None
    actual_position_pct: float | None = None
    plan_max_pct: float | None = None
    off_plan_count: int = 0
    off_plan_symbols: list[str] = Field(default_factory=list)
    active_plan_date: str = ""
    plan_symbols: list[PlanSymbolStatus] = Field(default_factory=list)


class StrategyBoardOut(BaseModel):
    config_key: str
    as_of: str | None = None
    source: str = "none"
    note: str = ""
    panel_symbols: list[str] = Field(default_factory=list)
    signals: list[StrategySignalRow] = Field(default_factory=list)
    positions: list[StrategyPositionRow] = Field(default_factory=list)
    risk_summary: RiskSummaryOut | None = None


class SignalPanelOut(BaseModel):
    symbols: list[str] = Field(default_factory=list)
    max_symbols: int = 10
    count: int = 0


class SignalPanelReplaceRequest(BaseModel):
    symbols: list[str] = Field(default_factory=list)


class SignalPanelMemberRequest(BaseModel):
    symbol: str = Field(description="支持 600519.SSE / SHSE.600519 / 600519")


class FinancialSnapshotOut(BaseModel):
    end_date: str
    revenue: float | None = None
    net_income: float | None = None
    revenue_yoy: float | None = None
    net_income_yoy: float | None = None
    roe: float | None = None
    debt_ratio: float | None = None


class FinancialSyncOut(BaseModel):
    last_sync_at: str
    latest_end_date: str = ""
    periods_count: int = 0
    sync_status: str = "ok"
    error_message: str = ""


class DisclosureOut(BaseModel):
    end_date: str
    pre_date: str = ""
    ann_date: str = ""
    actual_date: str = ""


class FundamentalsOut(BaseModel):
    vt_symbol: str
    ts_code: str
    snapshot: FinancialSnapshotOut | None = None
    sync: FinancialSyncOut | None = None
    disclosures: list[DisclosureOut] = Field(default_factory=list)



class PositionOut(BaseModel):
    symbol: str
    exchange: str
    vt_symbol: str
    cost_price: float
    volume: int
    buy_date: str
    notes: str = ""
    source: str = "manual"
    plan_pct: float | None = None
    sort_order: int = 0
    created_at: str = ""
    updated_at: str = ""


class PositionUpsertRequest(BaseModel):
    symbol: str = Field(description="支持 600519.SSE / SHSE.600519 / 600519")
    exchange: str | None = None
    cost_price: float = Field(gt=0)
    volume: int = Field(gt=0, description="须为 100 股整手")
    buy_date: str = Field(description="YYYY-MM-DD")
    notes: str = ""
    plan_pct: float | None = None


class TradingRiskPrefsOut(BaseModel):
    total_capital: float | None = None
    stop_loss_pct: float = 0.05
    caution_float_pct: float = -5.0
    realized_pnl_today: float | None = None


class TradingRiskPrefsPut(BaseModel):
    total_capital: float | None = None
    stop_loss_pct: float | None = None
    caution_float_pct: float | None = None
    realized_pnl_today: float | None = None


class NotifyLogItem(BaseModel):
    id: str
    event_type: str
    channel: str
    status: str
    error: str = ""
    created_at: str
    payload: dict[str, Any] = Field(default_factory=dict)


class NotifyLogOut(BaseModel):
    items: list[NotifyLogItem] = Field(default_factory=list)
    limit: int = 50
    count: int = 0

File: frontend/src/api/watchlist.ts
import { api } from './client'

export type WatchlistItem = {
  symbol: string
  exchange: string
  name: string
  industry?: string
  sort_order: number
  vt_symbol: string
  tf_symbol: string
  last_price: number | null
  change_pct: number | null
  turnover_rate: number | null
  volume: number | null
  amount: number | null
  volume_ratio: number | null
  suspended?: boolean
}

export type WatchlistGroup = {
  id: string
  name: string
  sort_order: number
}

export type GroupMembersBatchResult = {
  ok: boolean
  action: 'add' | 'remove'
  added: number
  removed: number
  skipped: number
  errors: Array<{ symbol: string; detail: string }>
}

export type Bar = {
  datetime: string
  open: number
  high: number
  low: number
  close: number
  volume: number
  turnover: number
}

export type BarsResponse = {
  symbol: string
  exchange: string
  vt_symbol: string
  interval: string
  bars: Bar[]
}

export type StrategySignalRow = {
  vt_symbol: string
  name: string
  last_price: number | null
  change_pct: number | null
  signal: string
  signal_label: string
  signal_date: string | null
  strength: number | null
  strength_tier?: string | null
  strength_tier_label?: string | null
  reason_summary: string
  ref_buy_price: number | null
  ref_sell_price: number | null
  ma_gap_pct: number | null
  bar_as_of: string | null
}

export type StrategyPositionRow = {
  vt_symbol: string
  name: string
  cost_price: number
  volume: number
  buy_date: string
  last_price: number | null
  market_value: number | null
  unrealized_pnl: number | null
  unrealized_pnl_pct: number | null
  t1_locked: boolean
  exit_signal: string
  exit_signal_label: string
  ref_sell_price: number | null
  reason_summary: string
  risk_tags?: string[]
  risk_primary?: string
}

export type RiskSummary = {
  total_capital: number | null
  actual_position_pct: number | null
}

export type TradingRiskPrefs = {
  total_capital: number | null
  stop_loss_pct: number
  caution_float_pct: number
  realized_pnl_today: number | null
}

export type TradingRiskPrefsPut = {
  total_capital?: number | null
  stop_loss_pct?: number | null
  caution_float_pct?: number | null
  realized_pnl_today?: number | null
}

export type StrategyBoard = {
  config_key: string
  signal_mode?: string
  as_of: string | null
  source: string
  note: string
  panel_symbols: string[]
  signals: StrategySignalRow[]
  positions: StrategyPositionRow[]
  risk_summary?: RiskSummary | null
}

export type SignalPanel = {
  symbols: string[]
  max_symbols: number
  count: number
}

export type PositionItem = {
  symbol: string
  exchange: string
  vt_symbol: string
  cost_price: number
  volume: number
  buy_date: string
  notes: string
  source: string
  sort_order: number
  created_at: string
  updated_at: string
}

export type PositionUpsert = {
  symbol: string
  exchange?: string
  cost_price: number
  volume: number
  buy_date: string
  notes?: string
}

export type NotifyLogItem = {
  id: string
  event_type: string
  channel: string
  status: string
  error: string
  created_at: string
  payload: Record<string, unknown>
}

export type NotifyLogOut = {
  items: NotifyLogItem[]
  limit: number
  count: number
}

export type Fundamentals = {
  vt_symbol: string
  ts_code: string
  snapshot: {
    end_date: string
    revenue: number | null
    net_income: number | null
    revenue_yoy: number | null
    net_income_yoy: number | null
    roe: number | null
    debt_ratio: number | null
  } | null
  sync: {
    last_sync_at: string
    latest_end_date: string
    periods_count: number
    sync_status: string
    error_message: string
  } | null
  disclosures: {
    end_date: string
    pre_date: string
    ann_date: string
    actual_date: string
  }[]
}

export type QuoteOut = {
  symbol: string
  exchange: string
  vt_symbol: string
  tf_symbol: string
  name: string
  last_price: number | null
  change_pct: number | null
  turnover_rate: number | null
  volume: number | null
  amount: number | null
  amplitude: number | null
  volume_ratio: number | null
  industry: string
}

export const watchlistApi = {
  list: (groupId?: string) => {
    const q = groupId ? `?group_id=${encodeURIComponent(groupId)}` : ''
    return api<WatchlistItem[]>(`/api/v1/watchlist${q}`)
  },
  strategyBoard: (opts?: { configKey?: string; signalMode?: string }) => {
    const params = new URLSearchParams()
    if (opts?.configKey) params.set('config_key', opts.configKey)
    if (opts?.signalMode) params.set('signal_mode', opts.signalMode)
    const q = params.toString() ? `?${params.toString()}` : ''
    return api<StrategyBoard>(`/api/v1/watchlist/strategy-board${q}`)
  },
  tradingRisk: () => api<TradingRiskPrefs>('/api/v1/watchlist/trading-risk'),
  putTradingRisk: (body: TradingRiskPrefsPut) =>
    api<TradingRiskPrefs>('/api/v1/watchlist/trading-risk', {
      method: 'PUT',
      body: JSON.stringify(body),
    }),
  notifyLog: (limit?: number) => {
    const q = limit != null ? `?limit=${encodeURIComponent(String(limit))}` : ''
    return api<NotifyLogOut>(`/api/v1/watchlist/notify-log${q}`)
  },
  signalPanel: () => api<SignalPanel>('/api/v1/watchlist/signal-panel'),
  replaceSignalPanel: (symbols: string[]) =>
    api<SignalPanel>('/api/v1/watchlist/signal-panel', {
      method: 'PUT',
      body: JSON.stringify({ symbols }),
    }),
  addSignalPanelMember: (symbol: string) =>
    api<SignalPanel>('/api/v1/watchlist/signal-panel/members', {
      method: 'POST',
      body: JSON.stringify({ symbol }),
    }),
  removeSignalPanelMember: (vtSymbol: string) =>
    api<SignalPanel>(`/api/v1/watchlist/signal-panel/members/${encodeURIComponent(vtSymbol)}`, {
      method: 'DELETE',
    }),
  listPositions: () => api<PositionItem[]>('/api/v1/watchlist/positions'),
  addPosition: (body: PositionUpsert) =>
    api<PositionItem>('/api/v1/watchlist/positions', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
  updatePosition: (vtSymbol: string, body: PositionUpsert) =>
    api<PositionItem>(`/api/v1/watchlist/positions/${encodeURIComponent(vtSymbol)}`, {
      method: 'PUT',
      body: JSON.stringify(body),
    }),
  removePosition: (vtSymbol: string) =>
    api<{ ok: boolean }>(`/api/v1/watchlist/positions/${encodeURIComponent(vtSymbol)}`, {
      method: 'DELETE',
    }),
  add: (symbol: string, name = '') =>
    api<WatchlistItem>('/api/v1/watchlist', {
      method: 'POST',
      body: JSON.stringify({ symbol, name }),
    }),
  remove: (vtSymbol: string) =>
    api<{ ok: boolean }>(`/api/v1/watchlist/${encodeURIComponent(vtSymbol)}`, {
      method: 'DELETE',
    }),
  groups: () => api<WatchlistGroup[]>('/api/v1/watchlist/groups'),
  createGroup: (name: string) =>
    api<WatchlistGroup>('/api/v1/watchlist/groups', {
      method: 'POST',
      body: JSON.stringify({ name }),
    }),
  deleteGroup: (id: string) =>
    api<{ ok: boolean }>(`/api/v1/watchlist/groups/${encodeURIComponent(id)}`, {
      method: 'DELETE',
    }),
  renameGroup: (id: string, name: string) =>
    api<WatchlistGroup>(`/api/v1/watchlist/groups/${encodeURIComponent(id)}`, {
      method: 'PATCH',
      body: JSON.stringify({ name }),
    }),
  reorderGroups: (groupIds: string[]) =>
    api<WatchlistGroup[]>('/api/v1/watchlist/groups/reorder', {
      method: 'PUT',
      body: JSON.stringify({ group_ids: groupIds }),
    }),
  addToGroup: (groupId: string, symbol: string) =>
    api<{ ok: boolean }>(`/api/v1/watchlist/groups/${encodeURIComponent(groupId)}/members`, {
      method: 'POST',
      body: JSON.stringify({ symbol }),
    }),
  removeFromGroup: (groupId: string, vtSymbol: string) =>
    api<{ ok: boolean }>(
      `/api/v1/watchlist/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(vtSymbol)}`,
      { method: 'DELETE' },
    ),
  batchGroupMembers: (groupId: string, symbols: string[], action: 'add' | 'remove') =>
    api<GroupMembersBatchResult>(
      `/api/v1/watchlist/groups/${encodeURIComponent(groupId)}/members/batch`,
      {
        method: 'POST',
        body: JSON.stringify({ symbols, action }),
      },
    ),
  bars: (vtSymbol: string, interval = 'd', limit = 120) =>
    api<BarsResponse>(
      `/api/v1/bars/${encodeURIComponent(vtSymbol)}?interval=${interval}&limit=${limit}`,
    ),
  quotes: (symbols: string) =>
    api<QuoteOut[]>(`/api/v1/quotes?symbols=${encodeURIComponent(symbols)}`),
  fundamentals: (vtSymbol: string) =>
    api<Fundamentals>(`/api/v1/watchlist/items/${encodeURIComponent(vtSymbol)}/fundamentals`),
}
